pkg/backend/openshiftcluster.go

```diff
@@ -130,13 +133,20 @@ func (ocb *openShiftClusterBackend) heartbeat(log *logrus.Entry, doc *api.OpenSh
 		for {
 			_, err := ocb.db.OpenShiftClusters.Lease(doc.Key)
 			if err != nil {
+				// if lost lease - cancel backend workers
+				// and revoke lease
+				if errors.Is(err, database.ErrorLostLease) {
+					cancel()
+					ocb.db.OpenShiftClusters.EndLease(doc.Key)
+				}
 				log.Error(err)
 				return
 			}
 
 			select {
 			case <-t.C:
 			case <-stop:
+				ocb.db.OpenShiftClusters.EndLease(doc.Key)
```

Comment:
We end lease each time when heartbeat is stopped. This way any `return` from `handle` will automatically cancel the lease. This way lease "tracking" is only in one place. We just update status when we required so in return, or soon after it. And `defer` will cancel the lease and backend workers.